Blocks: Tag Cloud

jahvi opened the pull request:
## Description
This is my attempt at implementing the tag cloud widget as a Gutenberg block, it's still a work in progress but I wanted to get some initial feedback while I'm at it.

Continued from #4994.
Fixes #1801.


## How has this been tested?
Tested in the latest master branch.

## Screenshots <!-- if applicable -->
![motion](https://user-images.githubusercontent.com/661330/42536345-ad3a6452-8489-11e8-99ff-52cd8d0629d2.gif)

## Types of changes
New feature: tag cloud block

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

jahvi committed:
Add tag cloud block

jahvi committed:
Add tag cloud block

jahvi commented:
Updated PR to match changes from #8287

Commit:
Fix phpcs format issues

Commit:
Fix phpcs format issues

jahvi commented:
Any update on this one? I rebased the latest changes from master but not sure if I should refactor anything else.

Soean commented:
@jahvi Can you rebase it again? I will test the block on my Gutenberg installation.

jahvi commented:
@Soean Sure thing, I just rebased it now.

Soean commented on packages/block-library/src/tag-cloud/index.js:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Internal dependencies
+ */
+import './editor.scss';
```
You don't have to include the editor.scss in index.js. There is now a editor.scss file in `/packages/block-libary/src/` where you can import the file.

jahvi replied:
My bad this should be fixed now.

Commit:
Move styles import to main editor.scss file

Commit:
Move styles import to main editor.scss file

jahvi committed:
Include server side script on load

jahvi commented:
Thanks for reopening, I just rebased and fixed some code to match the latest Gutenberg changes. I'm not sure how to fix the CI issues, can you guys advise?

Soean commented:
Nice work! You have to delete the generated test files and regenerate them.

Delete: 
- `core__tag-cloud__showTagCounts.json`
- `core__tag-cloud__showTagCounts.parsed.json`
- `core__tag-cloud__showTagCounts.serialized.html`
- `core__tag-cloud.json`
- `core__tag-cloud__showTagCounts.parsed.json`
- `core__tag-cloud__showTagCounts.serialized.html`

Run: `GENERATE_MISSING_FIXTURES=y npm run test-unit test/integration/full-content/full-content.spec.js`

jahvi committed:
Add missing fixtures

jahvi commented:
Thanks it should be fixed now, there are some other failing tests still but not sure if they're related to this PR.

Soean commented:
@jahvi Can you rebase it again? Then the tests should be correct.
The Gutenberg phase 2 with `Widgets to blocks` has startet, so I asked in the Core-editor chat for reviews of this PR: https://make.wordpress.org/core/2019/01/17/editor-chat-summary-january-16/

Commit:
Add missing fixture files

jahvi commented:
@Soean Rebased again, all tests are green now.

jahvi commented:
@noisysocks That's odd, I'll have a look!

jahvi commented:
@noisysocks I haven't been able to replicate the issue on my end, can you confirm what version of WP you're using?

![motion](https://user-images.githubusercontent.com/661330/51708086-e2ee0900-201a-11e9-9d03-3bfd300fbc7c.gif)

noisysocks commented:
I'm using [Gutenberg's built-in local environment](https://github.com/WordPress/gutenberg/blob/master/CONTRIBUTING.md#local-environment).

noisysocks commented:
It looks like `/block-renderer` is returning a 404 status because the API is not able to see that the `core/tag-cloud` block is registered on the server. 

Adding the PHP file which registers the block to `lib/load.php` fixes this.

```diff
diff --git a/lib/load.php b/lib/load.php
index ee1c973f9..da1a0f709 100644
--- a/lib/load.php
+++ b/lib/load.php
@@ -74,3 +74,6 @@ if ( ! function_exists( 'render_block_core_latest_posts' ) ) {
 if ( ! function_exists( 'render_block_core_shortcode' ) ) {
        require dirname( __FILE__ ) . '/../packages/block-library/src/shortcode/index.php';
 }
+if ( ! function_exists( 'render_block_core_tag_cloud' ) ) {
+       require dirname( __FILE__ ) . '/../packages/block-library/src/tag-cloud/index.php';
+}
```

noisysocks commented on packages/block-library/src/tag-cloud/index.php:
```diff
@@ -0,0 +1,61 @@
+
+	$args = array(
+		'echo'       => false,
+		'taxonomy'   => $attributes['taxonomy'],
```
This will cause an 'Undefined index' warning when rendering a Tag Cloud block that has no taxonomy set and PHP is set to display notices. You can test this by inserting a temporary `error_reporting( E_ALL );` statement at the top of the file.

Bailing out early when `$attributes['taxonomy']` is not set is probably the best bet.

```php
if ( ! isset( $attributes['taxonomy'] ) ) {
	return '';
}
```

noisysocks commented on packages/block-library/src/tag-cloud/index.php:
```diff
@@ -0,0 +1,61 @@
+	$args = array(
+		'echo'       => false,
+		'taxonomy'   => $attributes['taxonomy'],
+		'show_count' => ! empty( $attributes['showTagCounts'] ),
```
nit: Because `showTagCounts` has a `'default'` of `false`, it's safe to use `$attributes['showTagCounts']` directly here.

```suggestion
		'show_count' => $attributes['showTagCounts'],
```

noisysocks commented on packages/block-library/src/tag-cloud/index.php:
```diff
@@ -0,0 +1,61 @@
+	$tag_cloud = wp_tag_cloud( $args );
+
+	$block_content = sprintf(
+		$wrapper_markup,
```
nit: I think this would read clearer if we inlined `$wrapper_markup`. It aids with readability when variables are defined near to where they are used.

```php
$block_content = sprintf(
	'<p class="%1$s">%2$s</p>',
	esc_attr( $class ),
	$tag_cloud
);
```

noisysocks commented on packages/block-library/src/index.js:
```diff
@@ -94,11 +95,8 @@ export const registerCoreBlocks = () => {
-	].forEach( ( block ) => {
-		if ( ! block ) {
-			return;
-		}
```
Why remove this?

noisysocks commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -0,0 +1,123 @@
+				value: '',
+			},
+			...taxonomies,
+		];
```
nit: We could perform this array concatenation in `getTaxonomies()`. This would shorten `render()` which should help with readability.

gziolo replied:
It can also be renamed it to `getTaxonomyOptions` to take into account the proposed change.

noisysocks reviewed (dismissed):
This is looking great! Thanks so much for working on this 😍

I've left a few comments which need to be addressed. 

I tested this locally by applying the patch in https://github.com/WordPress/gutenberg/pull/7875#issuecomment-457434025 and it's looking really good! 

I'd like us to re-evaluate whether displaying a placeholder to select which taxonomy is shown is the best UX. My inclination is the the majority of users will want to display 'Tags' in the cloud instead of 'Categories'. Users can change the taxonomy to 'Categories' via the sidebar if appropriate. This looks to be how the Tag Cloud block was designed per https://github.com/WordPress/gutenberg/issues/1801.

cc. @melchoyce

jahvi commented:
@noisysocks Great I'll go through the changes then.

>> Adding the PHP file which registers the block to lib/load.php fixes this.

That's weird I'm pretty sure I did this, maybe I forgot to commit it. I'll make sure it's fix on the next update.

jahvi committed:
Merge branch 'add/block-tag-cloud' of github.com:jahvi/gutenberg into add/block-tag-cloud

jahvi committed:
Misc fixes to tag cloud widget

jahvi committed:
Update tag cloud fixtures

jahvi commented:
Thanks for the feedback! I went ahead and made the recommended changes, let me know if there's anything else I can do.

gziolo commented on packages/block-library/src/tag-cloud/index.php:
```diff
@@ -0,0 +1,63 @@
+
+	$tag_cloud = wp_tag_cloud( $args );
+
+	$block_content = sprintf(
```
Nit: this variable is obsolete, you can return here.

gziolo commented:
@mapk and @melchoyce - code wise this PR is looking great, how about design aspects? Do you think it's ready to go?

Commit:
Remove extra variable from render function

melchoyce commented:
> I'd like us to re-evaluate whether displaying a placeholder to select which taxonomy is shown is the best UX. My inclination is the the majority of users will want to display 'Tags' in the cloud instead of 'Categories'. Users can change the taxonomy to 'Categories' via the sidebar if appropriate. This looks to be how the Tag Cloud block was designed per #1801.

Agreed.

Likewise, I think we should rename this label: 

> Show tag counts

"Tag counts" doesn't make sense if you're displaying categories. I know this comes from the original widget, but I think it's an opportunity to add some clarity. In my original mockups I used "show post counts," which I think makes more sense — since you're counting the number of posts that use these categories and tags.

Soean reviewed (dismissed):
The block works great, but I found two additional points which we should solve before we merge.

- It is not possible to add a custom CSS class. Therefore you have do support the attribute `className` in the PHP file.
- If you don't have any tags, the block is not selectable anymore.

jahvi commented:
Working on the changes now.

> If you don't have any tags, the block is not selectable anymore.

What should I do in this case? Display a generic "no tags found" text?

Soean commented:
We have no error/success in the server side rendered blocks yet, so we should only show a message. In the comments block, we have a 'No comments to show.' text.
![bildschirmfoto 2019-01-30 um 14 09 47](https://user-images.githubusercontent.com/695201/51983349-bd3c8600-2498-11e9-902a-9b21a1926d75.png)

afercia commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -0,0 +1,117 @@
+						<SelectControl
+							options={ taxonomyOptions }
+							onChange={ this.setTaxonomy }
+						/>
```
This select element misses a properly associated `<label>` element. Should use a prop
`label={ __( 'Select a Taxonomy' ) }` 
And the `instructions` prop should be removed from the placeholder.

afercia commented:
Noticed a couple of things:

1
This select should have a properly associated `<label>` element:

<img width="709" alt="screenshot 2019-01-30 at 17 37 53" src="https://user-images.githubusercontent.com/1682452/51996875-eb7c8e80-24b5-11e9-803d-c88c17efdea4.png">

Noting that at this point, also other blocks with a placeholder should have a visible `<label>` element, for example:

<img width="716" alt="screenshot 2019-01-30 at 17 45 20" src="https://user-images.githubusercontent.com/1682452/51997451-07ccfb00-24b7-11e9-9608-962629783e98.png">

2
Once the tag cloud is added, there's no way to edit it (for example to change taxonomy) other than jumping to the sidebar. For some users, going to the sidebar is a terribly difficult task.

Other blocks have an "Edit" button in their toolbar that switches back the block to the placeholder view, for example: embeds.

Seems to me all the widget-blocks that have a placeholder view, should have an Edit button in the toolbar.

afercia commented:
Minor but worth considering: is it OK the tags are wrapped in a `<p>` paragraph? The classic tag cloud widget uses a `<div>`. Should usages outside of the post content (widget-block in the sidebar) be considered?

melchoyce commented:
Good call re: edit button. I kinda hate that pattern, but I think it makes sense to add in for consistency (and hopefully find a better solution in the future). 

re: div/p — are the classes the same? Thinking about how themes style widgets and how we can make it easy to extend those theme styles to the block.

afercia commented:
@melchoyce I don't think the CSS classes are the same: the one used here is `wp-block-tag-cloud`, which is new.

melchoyce commented:
Okay, in that case, it'll need to be uniquely styled regardless. 

Which do you think is more semantic both inside and outside `the_content` — div or p?

jahvi committed:
Misc a11y changes to tag cloud widget

jahvi committed:
Merge branch 'add/block-tag-cloud' of github.com:jahvi/gutenberg into add/block-tag-cloud

jahvi commented:
Just updated the PR with the suggested changes, in summary:

- Renamed "show tag counts" label
- Added `className` support
- Added empty message when no taxonomies are available
- Added labels to select fields
- Made "tags" the default taxonomy
- Added "Edit" icon to toolbar

Let me know if I should change anything else.

![motion](https://user-images.githubusercontent.com/661330/52231602-4c3f0900-28b2-11e9-8fab-d2dfd626519a.gif)

melchoyce commented:
I don't know that we lost any settings when converting core widgets to blocks — what if we didn't include them here, and only used it to display third-party widgets?

melchoyce commented:
I get "No taxonomies to show" when I add the block:

![image](https://user-images.githubusercontent.com/2846578/52234271-3ff58480-288f-11e9-89fb-4bddb91c4ef7.png)

But I do have tags and categories on my site, and they show up just fine if I manually select them in my sidebar.

jahvi committed:
Change tag taxonomy name

jahvi commented:
My bad I was using the wrong taxonomy name for tags, should be fixed now.

mapk commented:
Just testing this!  I like that there's no placeholder initially now. It's looking great! However when I click the edit button, I can't get back to see the preview mode. I even tried reselecting the tag option in the dropdown, but it didn't work.

![tagcloud](https://user-images.githubusercontent.com/617986/52312617-1a847b80-2960-11e9-8226-1bd174af6feb.gif)

melchoyce commented:
Instead of having the placeholder when you click edit, what if we removed edit and instead, Category was a transform option?

melchoyce commented:
This is what I'm thinking:

![image](https://user-images.githubusercontent.com/2846578/52376946-7408ab00-2a31-11e9-8ba3-6e4b9365814f.png)

mapk commented:
@melchoyce Do you think they would both appear in the block inserter as two different blocks at that point?

Another idea is to completely rename the block to "Word Cloud," default the block to tags, but offer a category option in the Inspector. This would keep it all under one block still. But this depends on "Word Cloud" being recognizable enough to work.

melchoyce commented:
When I hear word cloud, I think I'm going to get to input the words and have it make the cloud. I don't think we should make additional blocks, so maybe transform isn't appropriate here 🤷‍♀️

jahvi commented:
@mapk I see, based on the gif looks like you're selecting the same taxonomy which won't trigger the `onChange` event indeed.

Should I make it so it goes back even when selecting the same taxonomy? Or should I keep the edit icon in the toolbar even when the placeholder is visible so it can be clicked to go back?

I tried to keep it consistent with the RSS block where the edit icon is hidden when the placeholder is visible.

mapk commented:
> I tried to keep it consistent with the RSS block where the edit icon is hidden when the placeholder is visible.

The reason the RSS block works this way is because there's an actual button in the setup screen that can take you back. In the case of the tag cloud, there isn't.

> Should I make it so it goes back even when selecting the same taxonomy? Or should I keep the edit icon in the toolbar even when the placeholder is visible so it can be clicked to go back?

I don't think keeping the Edit icon there to click again for a preview is intuitive. BUT I really like the Edit icon to take me to the setup screen where I can select between "tags" or "categories". 

So we've got a few options here:
1. Include a "Preview" button next to the Edit icon similarly to the HTML block.

![html-block](https://user-images.githubusercontent.com/617986/52599856-ae49c200-2e0e-11e9-8802-f9515b9224d1.gif)

2. Or add a button to the setup screen so that when they make their selection, they also need to click a "submit" button to bring them to the Preview mode.

How do either of those sound?

mapk commented:
I guess the third option here could be to default the block to "tags" and just get rid of the setup screen completely. Only allow the "tag" or "category" option in the Inspector kind of how you have that there now.

aduth commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -0,0 +1,141 @@
+
+export default withSelect( ( select ) => {
+	return {
+		taxonomies: select( 'core' ).getTaxonomies( { per_page: -1 } ),
```
Is `per_page` a valid argument for fetching taxonomies, and if so, is `-1` a valid value?

noisysocks replied:
Yes, this query param will be ignored by the REST API. 

```suggestion
		taxonomies: select( 'core' ).getTaxonomies(),
```

noisysocks commented:
> I guess the third option here could be to default the block to "tags" and just get rid of the setup screen completely. Only allow the "tag" or "category" option in the Inspector kind of how you have that there now.

Personally I think this makes the most sense, see https://github.com/WordPress/gutenberg/pull/7875#pullrequestreview-196340095.

youknowriad commented:
Can we clarify what's remaining here? Can we get this in a shippable state for tomorrow? Happy to help with anything.

noisysocks committed:
Remove extra parameters in taxonomies query

Co-Authored-By: jahvi <javiervd@gmail.com>

jahvi committed:
Merge branch 'master' into add/block-tag-cloud

jahvi commented:
Should I go for the third option then? Remove the setup view and just allow for changing taxonomies on the sidebar?

It's a simple change so I can have it ready later today.

melchoyce commented:
Yeah, let's do that. Thanks!

jahvi committed:
Remove placeholder view from tag cloud

jahvi commented:
Removed the placeholder view now, I also changed the text when no tags are found to "No terms to show" since the previous "No taxonomies to show" didn't make much sense, hope that's ok.

![motion](https://user-images.githubusercontent.com/661330/52817521-faf5fd00-309b-11e9-9482-3b5e238ff35d.gif)

mapk reviewed (approved):
This looks great! I think we can :shipit: 

**Aside as another PR to improve this:**
One struggle I had was when I would select the block, many times I'd click on a link accidentally and the page would either reload to that link destination, or alert me that I'm leaving. I noticed that on the Latest Posts block, if I accidentally clicked a link, it would open in a new tab instead of reloading my current view. This worked better because I really just wanted to select the block. Maybe we can adopt that strategy for this in another PR?

gziolo reviewed (approved):
Awesome work @jahvi 💯

Code wise this PR is ready to land. When testing I noticed some very subtle visual differences between this block and the existing widget:

Block with tags selected:
![screen shot 2019-02-15 at 09 31 21](https://user-images.githubusercontent.com/699132/52844280-84d9b080-3104-11e9-90a7-edf1a400ff96.png)

Widget with tags selected:
![screen shot 2019-02-15 at 09 32 06](https://user-images.githubusercontent.com/699132/52844313-9a4eda80-3104-11e9-81aa-b8a99a31dfc6.png)

Block with categories selected:
![screen shot 2019-02-15 at 09 26 56](https://user-images.githubusercontent.com/699132/52844197-56f46c00-3104-11e9-9706-0090e53634dc.png)

Widget with categories selected:
![screen shot 2019-02-15 at 09 27 01](https://user-images.githubusercontent.com/699132/52844200-56f46c00-3104-11e9-85cc-a0b78fa55bf4.png)

I have no idea if this is something that needs to be fixed or is fine to proceed as is. I'll leave the final word to @melchoyce and @mapk.

There is also third option available to Tag Cloud widget which I can't see on the block. Not sure what `Link Categories` is in the first place but it would be worth investigating.

![screen shot 2019-02-15 at 09 34 36](https://user-images.githubusercontent.com/699132/52844597-324cc400-3105-11e9-8f32-6cf5a2e84da8.png)

gziolo commented:
One more thing which I want to confirm. Tag cloud widget is using wrapping `div` with `tagcloud` class name. Block uses wrapping `p` without this class. Is that expected? I personally don't see it as an issue, but maybe we want to stay closer to widgets in terms of syntax. This is how widgets look like:

![screen shot 2019-02-15 at 10 45 23](https://user-images.githubusercontent.com/699132/52848548-51505380-310f-11e9-8ecc-72012c9038c0.png)

melchoyce commented:
Is the difference just the underlines? We should keep those.

melchoyce commented:
Can we also make the error messages change based on which taxonomy is chosen? 

> No categories to show.
> No tags to show.

youknowriad commented:
I tested this and I feel it's good enough to land to include in Gutenberg 5.1. Follow-up improvements to styling, and messages can happen separately.

youknowriad commented:
Thanks everyone for the collaboration and @jahvi Awesome work

jahvi commented:
Thanks all for your help, I'll be glad to keep helping improve it.
